### PR TITLE
[ie/pr0gramm] Rewrite pr0gramm extractor

### PR DESCRIPTION
**Pull request by Grub4K**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Rewrite the extractor to improve on some low quality code. We now use the API and therefore have proper handling for auth-only media (using cookies).

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🗑️🛠️🧪

This pull request refactors and improves the `Pr0grammIE` extractor for the pr0gramm.com website. It adds support for more URL formats and metadata fields, and updates the tests accordingly. It also removes an unused import from `._extractors.py`.

> _`Pr0grammIE` grows_
> _More URLs and data_
> _Import no longer_

### Walkthrough
*  Refactor and improve `Pr0grammIE` extractor by adding support for more URL formats, fetching more metadata, handling errors and flags, and removing the dependency on `Pr0grammStaticIE` extractor ([link](https://github.com/yt-dlp/yt-dlp/pull/8151/files?diff=unified&w=0#diff-308fe09d0ad5be634e4df70963625b2cff808c98f9aed58711947ac2d105879bL1-R14), [link](https://github.com/yt-dlp/yt-dlp/pull/8151/files?diff=unified&w=0#diff-780b22dc7eb280f5a7b2bbf79aff17826de88ddcbf2fc1116ba19901827aa4e3L1522-R1522))
* Update and add tests for `Pr0grammIE` extractor to increase coverage and accuracy, and to ensure compatibility with different URL formats and metadata ([link](https://github.com/yt-dlp/yt-dlp/pull/8151/files?diff=unified&w=0#diff-308fe09d0ad5be634e4df70963625b2cff808c98f9aed58711947ac2d105879bL19-R130))

</details>

### CODE REVIEW

File: yt_dlp/extractor/pr0gramm.py [hunk 1]
```diff
@@ -1,97 +1,153 @@
...
+        'url': 'https://pr0gramm.com/user/froschler/dafur-ist-man-hier/5091290',
+        'only_matching': True,
+    }]
 
-        # Fetch author
-        uploader = self._html_search_regex(r'by\W+([\w-]+)\W+', webpage, 'uploader')
+    BASE_URL = 'https://pr0gramm.com'
+    API_URL = 'https://pr0gramm.com/api/items'
+    VIDEO_URL = 'https://img.pr0gramm.com'
+    THUMB_URL = 'https://thumb.pr0gramm.com'
```
Comment (pukkandan) [inline yt_dlp/extractor/pr0gramm.py:78]: Single use vars. Not needed

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/pr0gramm.py [hunk 2]
```diff
@@ -1,97 +1,150 @@
...
-        }, media_info)
+        return flags
 
+    def _call_api(self, endpoint, video_id, query={}, note='Downloading API json'):
+        data = self._download_json(f'https://pr0gramm.com/api/items/{endpoint}', video_id, query=query, note=note)
+        error = traverse_obj(data, ('error', {str}))
+        if error:
+            if error in ('nsfwRequired', 'nsflRequired', 'nsfpRequired'):
+                if not self._is_logged_in:
+                    self.raise_login_required(method='cookies')
```
Comment (bashonly) [inline yt_dlp/extractor/pr0gramm.py:103]:
```suggestion
                    self.raise_login_required()
```

**Review by bashonly — APPROVED** [review]

File: yt_dlp/extractor/pr0gramm.py [hunk 3]
```diff
@@ -1,97 +1,151 @@
...
-        return merge_dicts({
-            'id': video_id,
-            'title': 'pr0gramm-%s%s' % (video_id, (' by ' + uploader) if uploader else ''),
-            'uploader': uploader,
-            'upload_date': uploadTimestr
-        }, media_info)
+        return flags
 
+    def _call_api(self, endpoint, video_id, query={}, note='Downloading API json'):
+        data = self._download_json(f'https://pr0gramm.com/api/items/{endpoint}', video_id, query=query, note=note)
```
Comment (bashonly) [inline yt_dlp/extractor/pr0gramm.py:98]:
```suggestion
        data = self._download_json(
            f'https://pr0gramm.com/api/items/{endpoint}', video_id, query=query,
            note=note, expected_status=403)
```

Comment (Grub4K) [reply yt_dlp/extractor/pr0gramm.py:98]: I think we should also use the provided `msg` attribute if present. Ill cook something up later

**Pull request merged**